README.md
# GAME-QT-PUZZLE-Blubbles
A small puzzle game written in Qt4 (Jawbreaker clone)

## LICENSE
* GNU GPL V2

## COMPILE TOOLS
* 
 
## AUTHORS
* 

## LINKS
* http://sourceforge.net/projects/blubbels/


Changelog
Blubbels 1.0:
	* Ported to Maemo 5
	* More animations
	* Added a .desktop file and improved icons

Blubbels 0.9:
	* Changed home page to sf.net URL
	* Added window/application icon
	* Generate dialogs dynamically
	* Cleaned up source
	* Moving mouse pointer over a single bubble would show points for last group selected
	* Settings were not stored when exiting Blubbels by closing the window
	* Correctly center bubbles inside GameWidget
	* Updated German translation
	* Animations should run more smoothly now

Blubbels 0.8:
	* Fixed home page URL
	* Added german translation

Blubbels 0.7:
	* Make sure 'blubb.wav' is found in /usr/share/games/blubbels/
	* Reformatted About dialog
	* Changed bubble colors for more contrast on CRT screens

Blubbels 0.6:
	* Delay on Linux while playing sound should be gone now
	* "Selected" bubbles now have a border around them


game.h
/*
    Blubbels
    Copyright (C) 2007-2010 Christian Pulvermacher

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#ifndef GAME_H
#define GAME_H

#include <QList>
#include <QPixmap>
#include <QWidget>

class QMouseEvent;
class QSound;
class QTimer;

struct Box {
	Box(int x = -1, int y = -1): x(x), y(y) { }
	bool operator==(const Box &o) const { return x == o.x && y == o.y; }

	int x,y;
};

struct Bubble {
	Bubble(short color = 0):
		color(color), xoffset(0), yoffset(0), xspeed(0), yspeed(0)
	{
		//used for animations, should be enough for gridsize <= 16
		static short i = 0;
		id = i++;
	}
	short color; // 0 means empty

	//for animations
	float xoffset, yoffset; //in negative x/y-direction, 0 if none - in grid units
	float xspeed, yspeed;
	short id;
};


class GameWidget : public QWidget {
	Q_OBJECT
public:
	GameWidget(QWidget* parent = 0);
public slots:
	void enableAnimations(bool on) { use_animations = on; }
	void enableSound(bool on);
	void restart();
	void undo();
signals:
	void enableUndo(bool on);
	void showPoints(int points = 0);
	void newScore(int score);
protected: //overrides
	void mouseMoveEvent(QMouseEvent *event);
	void mouseReleaseEvent(QMouseEvent *event);
	void paintEvent(QPaintEvent *event);
	void resizeEvent(QResizeEvent *event);
private slots:
	void animationStep();
private:
	void checkGameOver();
	void compressBubbles();
	void getConnectedBubbles(Box pos, QList<Box> &list);
	Box getGridPos(int x, int y);

	const static int grid_size = 11;
	const static int num_colors = 5;

	Bubble grid[grid_size][grid_size];
	Bubble old_grid[grid_size][grid_size];
	QList<Box> selection;
	int bubblesize, margin, xoff, yoff;
	int score, oldscore;
	bool use_animations, use_sound;
	QPixmap bubblePixmap[num_colors+1]; // position 0 is empty

	QSound *pop;
	QTimer *timer; //for animations
};
#endif


main.h
/*
    Blubbels
    Copyright (C) 2007-2009 Christian Pulvermacher

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#ifndef MAIN_H
#define MAIN_H

#include <QMainWindow>
#include <QSettings>

class GameWidget;
class QAction;
class QCloseEvent;
class QLabel;

class MainWindow : public QMainWindow {
	Q_OBJECT
public:
	MainWindow(QWidget *parent = 0);
protected:
	void closeEvent(QCloseEvent *ev);
private slots:
	void about();
	void help();
	void setScore(int score);
	void showPoints(int points);
	void showStatistics();
private:
	GameWidget *gamewidget;
	QLabel *score_label;
	QSettings settings;
	QAction *use_animation;
	QAction *use_sound;
};
#endif


main.cpp
/*
    Blubbels
    Copyright (C) 2007-2010 Christian Pulvermacher

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#include "game.h"
#include "main.h"

#include <QAction>
#include <QApplication>
#include <QCoreApplication>
#include <QLabel>
#include <QLocale>
#include <QMenu>
#include <QMenuBar>
#include <QMessageBox>
#include <QStatusBar>
#include <QToolBar>
#include <QTranslator>

#define MAEMO (defined Q_WS_HILDON || defined Q_WS_MAEMO_5)

//default settings
const bool USE_ANIMATIONS = true;
#if MAEMO
const bool USE_SOUND = false;
#else
const bool USE_SOUND = true;
#endif


MainWindow::MainWindow(QWidget *parent) :
	QMainWindow(parent),
	gamewidget(new GameWidget(this)),
	score_label(new QLabel(this))
{
	setWindowTitle("Blubbels");

	gamewidget->enableAnimations(settings.value("use_animation", USE_ANIMATIONS).toBool());
	gamewidget->enableSound(settings.value("use_sound", USE_SOUND).toBool());
	setCentralWidget(gamewidget);

#ifndef USE_SHARED_DIR
	setWindowIcon(QIcon("icons/64x64/blubbels.png"));
#else
	setWindowIcon(QIcon("/usr/share/icons/hicolor/64x64/apps/blubbels.png"));
#endif

	//construct actions
	QAction *new_game = new QAction(QIcon(":/icons/new.png"), tr("&New Game"), this);
	new_game->setShortcut(tr("Ctrl+N"));
	QAction *undo = new QAction(QIcon(":/icons/undo.png"), tr("&Undo"), this);
	undo->setShortcut(tr("Ctrl+Z"));
	undo->setEnabled(false);
	use_sound = new QAction(tr("&Use Sound"), this);
	use_sound->setCheckable(true);
	use_sound->setChecked(settings.value("use_sound", USE_SOUND).toBool());
	use_animation = new QAction(tr("&Animate Bubbles"), this);
	use_animation->setCheckable(true);
	use_animation->setChecked(settings.value("use_animation", USE_ANIMATIONS).toBool());

#if !MAEMO
	statusBar()->addPermanentWidget(score_label);

	//create toolbar
	QToolBar *toolbar = new QToolBar(tr("Show Toolbar"), this);
	toolbar->setObjectName("toolBar"); //used by saveState()
	toolbar->addAction(new_game);
	toolbar->addSeparator();
	toolbar->addAction(undo);
	addToolBar(toolbar);

	//create menu
	QMenu *game_menu = menuBar()->addMenu(tr("&Game"));
	game_menu->addAction(new_game);
	game_menu->addSeparator();
	game_menu->addAction(tr("&Statistics"), this, SLOT(showStatistics()), tr("Ctrl+S"));
	game_menu->addSeparator();
	game_menu->addAction(tr("&Quit"), this, SLOT(close()), tr("Ctrl+Q"));

	QMenu *edit_menu = menuBar()->addMenu(tr("&Edit"));
	edit_menu->addAction(undo);

	QMenu *settings_menu = menuBar()->addMenu(tr("&Settings"));
	settings_menu->addAction(use_sound);
	settings_menu->addAction(use_animation);
	settings_menu->addSeparator();
	settings_menu->addAction(toolbar->toggleViewAction());

	QMenu *help_menu = menuBar()->addMenu(tr("&Help"));
	help_menu->addAction(tr("&How to Play"), this, SLOT(help()), tr("F1"));
	help_menu->addSeparator();
	help_menu->addAction(tr("&About"), this, SLOT(about()));
	help_menu->addAction(tr("About &Qt"), qApp, SLOT(aboutQt()));
#else
	//maemo only supports a limited number of menu items
	//TODO: correct for Qt 4.6?
	menuBar()->addAction(tr("&Statistics"), this, SLOT(showStatistics()));
	menuBar()->addAction(tr("&About"), this, SLOT(about())); menuBar()->addAction(new_game);
	menuBar()->addAction(undo); //not visible when disabled, so we'll move it to the end to avoid confusion
#endif

	connect(new_game, SIGNAL(triggered()),
		gamewidget, SLOT(restart()));
	connect(undo, SIGNAL(triggered()),
		gamewidget, SLOT(undo()));
	connect(use_animation, SIGNAL(toggled(bool)),
		gamewidget, SLOT(enableAnimations(bool)));
	connect(use_sound, SIGNAL(toggled(bool)),
		gamewidget, SLOT(enableSound(bool)));

	connect(gamewidget, SIGNAL(enableUndo(bool)),
		undo, SLOT(setEnabled(bool)));
	connect(gamewidget, SIGNAL(showPoints(int)),
		this, SLOT(showPoints(int)));
	connect(gamewidget, SIGNAL(newScore(int)),
		this, SLOT(setScore(int)));

	restoreState(settings.value("window_state", saveState()).toByteArray());
	resize(settings.value("window_size", size()).toSize());
	move(settings.value("window_pos", pos()).toPoint());
}


void MainWindow::about() {
	QMessageBox::about(this, tr("About Blubbels"),
		tr("<center><h1>Blubbels 1.0</h1>\
A Jawbreaker&trade; clone written in Qt 4\
<p>The current version of Blubbels can be found on <a href=\"http://sourceforge.net/projects/blubbels/\">http://sourceforge.net/projects/blubbels/</a>.</p>\
<small><p>&copy;2007-2010 Christian Pulvermacher &lt;pulvermacher@gmx.de&gt;<br>German Translation: &copy;2009 Dominic Hopf &lt;dh@dmaphy.de&gt;</p></small></center>\
%1")
#if !MAEMO
	.arg("<p>This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.</p> <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.</p> <p>You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.</p>")
#else
	.arg("<p>This program is free software; License: <a href=\"http://www.gnu.org/licenses/gpl-2.0.html\">GNU GPL 2</a> or later.</p>")
#endif
	);
}


void MainWindow::help()
{
#if !MAEMO
	QMessageBox::about(this, tr("How to Play"),
		tr("<b>How to play Blubbels</b><br>\
<p>In the main window you can see lots of colored bubbles, arranged in 'stacks'.</p>\
<p>If there are two or more bubbles of the same color next to each other (either horizontally or vertically) they can be removed. You can see the points that will be awarded for doing so by hovering your mouse cursor over the bubbles.</p>\
<p>Remove the group of bubbles by clicking on them. Any bubbles that rest above these will then fall down.  After completely emptying a stack, a new stack of bubbles of random height will be moved in from the left.</p>\
<p>The game ends if there are no bubbles left on the gameboard that can be removed.</p>"));
#endif
}


//window closed, save settings
void MainWindow::closeEvent(QCloseEvent* /*ev*/)
{
	//state of use_sound is saved by GameWidget::enableSound()
	settings.setValue("use_animation", use_animation->isChecked());
	settings.setValue("window_state", saveState());
	settings.setValue("window_size", size());
	settings.setValue("window_pos", pos());
	settings.sync();
}


//display current score in status bar (independent of other messages)
void MainWindow::setScore(int score)
{
#if MAEMO
	setWindowTitle(tr("Blubbels - Score: %n", "", score));
#else
	score_label->setText(tr("Score: %n", "", score));
#endif
}


void MainWindow::showPoints(int points)
{
#if !MAEMO
	if(points)
		statusBar()->showMessage(tr("%1 Points").arg(points));
	else
		statusBar()->clearMessage();
#endif
}


void MainWindow::showStatistics()
{
	QMessageBox::information(this, tr("Statistics"),
	tr("<table>\
<tr><td>Average:</td><td align='right'>%1</td></tr>\
<tr><td>Highscore:</td><td align='right'>%2</td></tr>\
<tr><td>Games played:</td><td align='right'>%3</td></tr>\
</table>")
	.arg(int(settings.value("average", 0).toDouble()))
	.arg(settings.value("highscore", 0).toInt())
	.arg(settings.value("games_played", 0).toInt()));
}


int main(int argc, char* argv[])
{
	QApplication app(argc, argv);

	QCoreApplication::setOrganizationName("Blubbels");
	QCoreApplication::setApplicationName("Blubbels");

	QTranslator translator;

#ifndef USE_SHARED_DIR
	//use pwd
	QString path = "";
#else
#if !MAEMO
	QString path = "/usr/share/games/blubbels/";
#else
	QString path = "/usr/share/blubbels/"; 
#endif
#endif
	translator.load(path + "blubbels_" + QLocale::system().name());
	app.installTranslator(&translator);

	MainWindow *mw = new MainWindow;
	mw->show();
	return app.exec();
}


game.cpp
/*
    Blubbels
    Copyright (C) 2007-2010 Christian Pulvermacher

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#include "game.h"

#include <QColor>
#include <QFileInfo>
#include <QMessageBox>
#include <QMouseEvent>
#include <QPainter>
#include <QPainterPath>
#include <QSettings>
#include <QSound>
#include <QTimer>

#include <cstdlib>


#define MAEMO (defined Q_WS_HILDON || defined Q_WS_MAEMO_5)


const QColor bubbleColor[] = {
	Qt::white, //"empty", shouldn't be shown
	Qt::green,
	Qt::blue,
	Qt::red,
	QColor(255, 180, 0),
	Qt::gray
};

const float initial_speed = 0.07;


GameWidget::GameWidget(QWidget* parent) :
	QWidget(parent),
	bubblesize(0),
	pop(0),
	timer(new QTimer(this))
{
	setMouseTracking(true);

	setAttribute(Qt::WA_OpaquePaintEvent);
	setMinimumSize(90, 90);

	connect(timer, SIGNAL(timeout()),
		this, SLOT(animationStep()));
	timer->setInterval(50);

	srand(time(0));
	restart();
}


//translates cursor position into grid coordinates
Box GameWidget::getGridPos(int x, int y)
{
	Box result; //invalid default value
	//make coords relative to bubbles
	x -= xoff; y -= yoff;

	if(x >=0 and y >= 0
	and (x-margin/2) / (bubblesize+margin) < grid_size and (y-margin/2) / (bubblesize+margin) < grid_size
	and grid[(x-margin/2)/(bubblesize+margin)][(y-margin/2)/(bubblesize+margin)].color) {
		result.x = (x-margin/2) / (bubblesize+margin);
		result.y = (y-margin/2) / (bubblesize+margin);
	}
	return result;
}


//advances all animations by one frame
void GameWidget::animationStep()
{
	bool animations_found = false;
	for(int x = 0; x < grid_size; x++) {
		for(int y = 0; y < grid_size; y++) {
			if(grid[x][y].yoffset) {
				grid[x][y].yoffset -= grid[x][y].yspeed;
				grid[x][y].yspeed *= 1.7;

				if((grid[x][y].yspeed > 0 and grid[x][y].yoffset <= 0)
				or (grid[x][y].yspeed < 0 and grid[x][y].yoffset >= 0))
					grid[x][y].yoffset = 0;
				else
					animations_found = true;
			}
			if(grid[x][y].xoffset) {
				grid[x][y].xoffset -= grid[x][y].xspeed;
				grid[x][y].xspeed *= 1.7;

				if((grid[x][y].xspeed > 0 and grid[x][y].xoffset <= 0)
				or (grid[x][y].xspeed < 0 and grid[x][y].xoffset >= 0))
					grid[x][y].xoffset = 0;
				else
					animations_found = true;
			}
		}
	}

	if(!animations_found)
		timer->stop();
	update();
}


//let bubbles fall down / move stacks to right
void GameWidget::compressBubbles()
{
	for(int x = grid_size-1; x >= 0; x--) {
		int hole = -1;
		int holes = 0;
		bool bubble_found = false;
		for(int y = 0; y <= grid_size; y++) {
			if(y == grid_size or grid[x][y].color) { //not empty
				if(holes) { //hole found, move bubbles down
					for(int i = hole; i > 0; i--) {
						grid[x][i] = grid[x][i-holes];

						if(use_animations) {
							grid[x][i].yoffset = holes;
							grid[x][i].yspeed = initial_speed;
						}
					}
					for(int i = 0; i < holes; i++)
						grid[x][i] = Bubble();
				}
				bubble_found = true;
				hole = -1;
				holes = 0;
			} else if(bubble_found) {
				hole = y;
				holes++;
			}
		}
	}

	for(int x = grid_size-1; x >= 0; x--) {
		if(!grid[x][grid_size-1].color) { //collumn empty
			//shift remaining collums
			int x2 = x;
			while(--x2 >= 0) {
				for(int y = 0; y < grid_size; y++) {
					grid[x2+1][y] = grid[x2][y];

					if(use_animations) {
						grid[x2+1][y].xoffset += 1;
						grid[x2+1][y].xspeed = initial_speed;
					}
				}
			}

			//add new col
			const int height = int(double(rand())/(RAND_MAX + 1.0)*grid_size);
			for(int y = grid_size-1; y >= height; y--) {
				grid[0][y] = Bubble(double(rand())/(RAND_MAX + 1.0)*num_colors+1);

				if(use_animations) {
					//place new col off screen
					grid[0][y].xoffset = (width() - grid_size*(bubblesize+margin))/bubblesize/2;
					grid[0][y].xspeed = initial_speed;
				}
			}

			//erase remains of old collumn
			for(int y = height-1; y >= 0; y--)
				grid[0][y] = Bubble();

			x++; //make sure to run over collumn again
		}
	}
	if(use_animations)
		timer->start();

	update();
	checkGameOver();
}


//returns list of all bubbles connected to pos
void GameWidget::getConnectedBubbles(Box pos, QList<Box> &list)
{
	int x = pos.x; int y = pos.y;
	const short color = grid[x][y].color;
	if(!color) //nothing to select, exit
		return;

	list << pos;

	x--; //left
	if(x >= 0 and color == grid[x][y].color and !list.contains(Box(x, y))) {
		getConnectedBubbles(Box(x, y), list);
	}
	x += 2; //right
	if(x < grid_size and color == grid[x][y].color and !list.contains(Box(x, y))) {
		getConnectedBubbles(Box(x, y), list);
	}
	x--; y--; //up
	if(y >= 0 and color == grid[x][y].color and !list.contains(Box(x, y))) {
		getConnectedBubbles(Box(x, y), list);
	}
	y += 2; //down
	if(y < grid_size and color == grid[x][y].color and !list.contains(Box(x, y))) {
		getConnectedBubbles(Box(x, y), list);
	}
}


void GameWidget::checkGameOver()
{
	for(int x = 0; x < grid_size; x++) {
		for(int y = 0; y < grid_size; y++) {
			QList<Box> connected;
			getConnectedBubbles(Box(x,y), connected);
			if(connected.size() > 1)
				return; //found cluster, continue game
		}
	}
	//assertion: no connected bubbles left in grid

	//update stats
	QSettings settings;
	settings.setValue("average",
		(settings.value("average", 0).toDouble()*settings.value("games_played", 0).toInt() + score) / (settings.value("games_played", 0).toInt()+1));
	settings.setValue("games_played", settings.value("games_played", 0).toInt()+1);

	if(score > settings.value("highscore", 0).toInt()) {
		QMessageBox::information(this, tr("New Highscore"), tr("Your Score: %n<br>Congratulations!", "", score) + tr("<br>You broke the previous record of %n points!", "", settings.value("highscore", 0).toInt()));
		settings.setValue("highscore", score);
	} else {
		QMessageBox::information(this, tr("Game Over"), tr("Your Score: %n<br>Congratulations!", "", score));
	}

	restart();
}


void GameWidget::mouseMoveEvent(QMouseEvent *event)
{
	if(event->buttons())
		return; //don't change selection while mouse button is down

	const Box hovered = getGridPos(event->x(), event->y());
	if(hovered.x == -1) { //nothing under cursor
		emit showPoints();
		selection.clear();
	} else if(!selection.contains(hovered)) { //cursor not in old selection
		selection.clear();
		getConnectedBubbles(hovered, selection);
		const int n = selection.size();
		if(n > 1) {
			emit showPoints(n*n);
		} else {
			emit showPoints();
			selection.clear(); //don't select single bubbles
		}
	}
	update();
}


void GameWidget::mouseReleaseEvent(QMouseEvent *event)
{
	if(selection.size() < 2 or !selection.contains(getGridPos(event->x(), event->y())))
		return; //nothing selected or mouse pointer not in selection

#if !MAEMO
	if(use_sound)
		pop->play(); //pop is initialized by enableSound()
#endif

	//backup
	oldscore = score;
	for(int x = 0; x < grid_size; x++)
		for(int y = 0; y < grid_size; y++)
			old_grid[x][y] = grid[x][y];

	//remove bubbles
	foreach(Box newpos, selection) {
		grid[newpos.x][newpos.y] = Bubble();
	}

	score += selection.size()*selection.size();
	selection.clear();

	emit enableUndo(true);
	emit newScore(score);
	emit showPoints();

	compressBubbles();

	mouseMoveEvent(event); //draw selection if there's a new set of connected bubbles
}


void GameWidget::resizeEvent(QResizeEvent* /*event*/)
{
	const int old_bubblesize = bubblesize;
	const int size = qMin(width(), height());
	bubblesize = 3.0*size/(4.0*grid_size+1); //best size for 1/3 margin
	margin = bubblesize/5; //separates bubbles from each other and the widget border

	const int round_error = size % int(grid_size*bubblesize + (grid_size+1)*margin);
	xoff = (width()-size+round_error)/2;
	yoff = (height()-size+round_error)/2;

	if(bubblesize != old_bubblesize) {
		//generate new pixmaps
		for(int i = 1; i <= num_colors; i++) {
			bubblePixmap[i] = QPixmap(bubblesize, bubblesize);
			bubblePixmap[i].fill(Qt::transparent);

			QPainter painter(&bubblePixmap[i]);
			painter.setPen(Qt::black);
			painter.setRenderHint(QPainter::Antialiasing, true);

			QRadialGradient radialGrad(QPointF(0.3*bubblesize, 0.3*bubblesize), 0.8*bubblesize);
			radialGrad.setColorAt(0, Qt::white);
			radialGrad.setColorAt(0.4, bubbleColor[i]);
			radialGrad.setColorAt(1, Qt::black);

			painter.setBrush(QBrush(radialGrad));
			painter.drawEllipse(0, 0, bubblesize, bubblesize);
		}
	}

	emit showPoints();
	update();
}


//enable sound and initialize sound system
void GameWidget::enableSound(bool on)
{
#if !MAEMO
	use_sound = on;

	QString path = "";
#ifdef USE_SHARED_DIR
	path = "/usr/share/games/blubbels/";
#endif
	const static QString soundfile = path + "blubb.wav";
	if(use_sound and !pop) {
		if(!QSound::isAvailable()) {
			use_sound = false;
			//TODO: use MessageBox
			qWarning(qPrintable(tr("Sound system not available, disabling sound!")));
		} else if(QFileInfo(soundfile).isReadable()) {
			pop = new QSound(soundfile);
		} else {
			use_sound = false; //disable sound for now
			qWarning(qPrintable(tr("Couldn't open %1, disabling sound!").arg(soundfile)));
		}
	}

	QSettings settings;
	settings.setValue("use_sound", use_sound);
#endif
}


//start new game, fill grid with bubbles
void GameWidget::restart()
{
	score = 0;
	emit newScore(score);
	emit showPoints();

	emit enableUndo(false);

	selection.clear();

	//seed field with random non-empty bubbles (values 1..num_colors)
	for(int x = 0; x < grid_size; x++) {
		for(int y = 0; y < grid_size; y++) {
			grid[x][y] = Bubble(double(rand())/(RAND_MAX+1.0)*num_colors+1);

			if(use_animations) {
				grid[x][y].xoffset = grid_size + (width() - grid_size*(bubblesize+margin))/bubblesize/2;
				grid[x][y].xspeed = initial_speed;
			}
		}
	}

	if(use_animations)
		timer->start();

	update();
}


//undo the last move
void GameWidget::undo()
{
	emit enableUndo(false);

	score = oldscore;
	emit newScore(score);
	selection.clear();
	emit showPoints();

	if(use_animations) {
		int shift = 0; //number of new collums in the action we want to undo
		for(int x = grid_size-1; x >= shift; x--) {
			bool not_found = false;
			for(int y = 1; y < grid_size; y++) {
				if(grid[x][y].id == old_grid[x-shift][y].id)
					break; //nothing changed

				not_found = true;
				for(int y_old = y-1; y_old >= 0; y_old--) { //search upward
					if(grid[x][y].id == old_grid[x-shift][y_old].id) {
						old_grid[x-shift][y_old].yoffset = y_old - y;
						old_grid[x-shift][y_old].yspeed = -initial_speed;

						not_found = false;
						break; //ids are unique
					}
				}

			}

			if(not_found) {//we couldn't find at least one bubble in it's collum
				shift++; //so it must have been the whole collumn that moved

			}

			if(shift > 0) {
				for(int y = 0; y < grid_size; y++) {
					old_grid[x-shift][y].xoffset = -shift;
					old_grid[x-shift][y].xspeed = -initial_speed;
				}
			}

			//don't skip collumns
			//TODO: only fixes special case shift=2
			if(not_found and shift > 1) {
				for(int y = 0; y < grid_size; y++) {
					old_grid[x-shift+1][y].xoffset = -shift;
					old_grid[x-shift+1][y].xspeed = -initial_speed;
				}
			}
		}
		
		timer->start();
	}

	for(int x = 0; x < grid_size; x++)
		for(int y = 0; y < grid_size; y++)
			grid[x][y] = old_grid[x][y];

	update();
}


void GameWidget::paintEvent(QPaintEvent* /*event*/)
{
	QPainter painter(this);
	painter.fillRect(0, 0, width(), height(), QBrush(Qt::black));
	painter.translate(xoff, yoff);

	for(int i = 0; i < grid_size; i++) {
		for(int j = 0; j < grid_size; j++) {
			if(!grid[i][j].color) //empty
				continue;

			QRectF target(
				(i-grid[i][j].xoffset) * (bubblesize+margin) + margin,
				(j-grid[i][j].yoffset) * (bubblesize+margin) + margin,
				bubblesize, bubblesize);
			QRectF source(0, 0, bubblesize, bubblesize);
			painter.drawPixmap(target, bubblePixmap[grid[i][j].color], source);
		}
	}

	//draw selection
	if(selection.empty() or selection.size() == 1)
		return;

	const short color = grid[selection.first().x][selection.first().y].color;
	painter.setPen(QPen(bubbleColor[color], margin/3, Qt::DotLine, Qt::RoundCap));
	QPainterPath path;

	foreach(Box b, selection) {
		if(b.x == 0 || grid[b.x-1][b.y].color != color) {
			QPointF point(b.x*(bubblesize+margin)+qreal(margin)/2, qreal(b.y*(bubblesize+margin)+margin/2));
			path.moveTo(point);
			point.ry() += bubblesize+margin;
			path.lineTo(point);

		}
		if(b.x == grid_size-1 || grid[b.x+1][b.y].color != color) {
			QPointF point(b.x*(bubblesize+margin)+qreal(margin)*3/2+bubblesize, b.y*(bubblesize+margin)+qreal(margin)/2);
			path.moveTo(point);
			point.ry() += bubblesize+margin;
			path.lineTo(point);
		}
		if(b.y == 0 || grid[b.x][b.y-1].color != color) {
			QPointF point(qreal(b.x*(bubblesize+margin)+margin/2), qreal(b.y*(bubblesize+margin)+margin/2));
			path.moveTo(point);
			point.rx() += bubblesize+margin;
			path.lineTo(point);
		}
		if(b.y == grid_size-1 || grid[b.x][b.y+1].color != color) {
			QPointF point(qreal(b.x*(bubblesize+margin)+margin/2), qreal(b.y*(bubblesize+margin)+margin*3/2+bubblesize));
			path.moveTo(point);
			point.rx() += bubblesize+margin;
			path.lineTo(point);
		}
	}
	painter.drawPath(path);
}
